akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -38,89 +40,173 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
 
   def flowController: IncomingFlowController = IncomingFlowController.default(settings)
 
-  private var incomingStreams = new immutable.TreeMap[Int, IncomingStreamState]
+  private var streamStates = new immutable.TreeMap[Int, StreamState]
   private var largestIncomingStreamId = 0
   private var outstandingConnectionLevelWindow = Http2Protocol.InitialWindowSize
   private var totalBufferedData = 0
 
-  private def streamFor(streamId: Int): IncomingStreamState =
-    incomingStreams.get(streamId) match {
+  /**
+   * The "last peer-initiated stream that was or might be processed on the sending endpoint in this connection"
+   *
+   * @see http://httpwg.org/specs/rfc7540.html#rfc.section.6.8
+   */
+  def lastStreamId(): Int = largestIncomingStreamId
+
+  private def streamFor(streamId: Int): StreamState =
+    streamStates.get(streamId) match {
       case Some(state) => state
       case None =>
         if (streamId <= largestIncomingStreamId) Closed // closed streams are never put into the map
         else if (isUpgraded && streamId == 1) {
+          require(isServer)
           // Stream 1 is implicitly "half-closed" from the client toward the server (see Section 5.1), since the request is completed as an HTTP/1.1 request
           // https://http2.github.io/http2-spec/#discover-http
           largestIncomingStreamId = streamId
-          incomingStreams += streamId -> HalfClosedRemote
-          HalfClosedRemote
+          streamStates += streamId -> HalfClosedRemoteWaitingForOutgoingStream(0)
+          HalfClosedRemoteWaitingForOutgoingStream(0)
         } else {
           largestIncomingStreamId = streamId
-          incomingStreams += streamId -> Idle
+          streamStates += streamId -> Idle // FIXME: is that needed at all? Should move out of Idle immediately
           Idle
         }
     }
-  def handleStreamEvent(e: StreamFrameEvent): Unit = {
+
+  /** Called by Http2ServerDemux to let the state machine handle StreamFrameEvents */
+  def handleStreamEvent(e: StreamFrameEvent): Unit =
     updateState(e.streamId, _.handle(e))
-  }
 
+  /** Called by Http2ServerDemux when a stream comes in from the user-handler */
   def handleOutgoingCreated(stream: Http2SubStream): Unit =
-    updateState(stream.streamId, _.handleOutgoingCreated(stream))
+    if (streamFor(stream.streamId) != Closed) {
```

Comment:
Since stream handling now has to deal with outgoing streams, we create those here and pass those to the state `handleOutgoingCreated`. There's another callback `handleOutgoingCreatedAndFinished` if the stream has no entity, so we can skip the logic for tracking sending entity data.